File: data/benchmark/src/benchmark/extraction/extractor.py
"""Vendor mention extraction engine — 3-tier extraction from transcript turns.

Port of packages/shared/src/extractor.ts
"""

from __future__ import annotations

import json
import re
from typing import Callable

from ..parsers.types import ParsedTurn
from .package_map import resolve_package_to_vendor as _resolve_static
from .types import VendorMention, VendorTaxonomy

# ── Package Manager Patterns ────────────────────────────────────────

INSTALL_PATTERNS = [
    re.compile(r"(?:npm\s+install|npm\s+i|npx)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)"),
    re.compile(r"(?:pnpm\s+add|pnpm\s+install)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)"),
    re.compile(r"(?:yarn\s+add)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)"),
    re.compile(r"(?:pip\s+install|pip3\s+install)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)"),
    re.compile(r"(?:brew\s+install)\s+([^\s|&;]+)"),
    re.compile(r"(?:docker\s+pull)\s+([^\s|&;]+)"),
]

# ── Connection String Patterns ──────────────────────────────────────

CONNECTION_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r"neon\.tech", re.I), "neon"),
    (re.compile(r"supabase\.co", re.I), "supabase"),
    (re.compile(r"supabase\.com", re.I), "supabase"),
    (re.compile(r"planetscale\.com", re.I), "planetscale"),
    (re.compile(r"turso\.io", re.I), "turso"),
    (re.compile(r"cockroachlabs\.cloud", re.I), "cockroachdb"),
    (re.compile(r"upstash\.io", re.I), "upstash"),
    (re.compile(r"xata\.sh", re.I), "xata"),
    (re.compile(r"fauna\.com", re.I), "fauna"),
    (re.compile(r"convex\.cloud", re.I), "convex"),
    (re.compile(r"railway\.app", re.I), "railway-postgres"),
    (re.compile(r"vercel\.app", re.I), "vercel-edge-functions"),
    (re.compile(r"fly\.dev", re.I), "fly-io"),
    (re.compile(r"workers\.dev", re.I), "cloudflare-workers"),
    (re.compile(r"deno\.dev", re.I), "deno-deploy"),
]

# ── ENV Var Patterns ────────────────────────────────────────────────

ENV_VAR_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r"SUPABASE_URL|SUPABASE_KEY|SUPABASE_ANON_KEY|NEXT_PUBLIC_SUPABASE", re.I), "supabase"),
    (re.compile(r"NEON_DATABASE_URL|NEON_DB_URL", re.I), "neon"),
    (re.compile(r"TURSO_DATABASE_URL|TURSO_AUTH_TOKEN", re.I), "turso"),
    (re.compile(r"PLANETSCALE_|DATABASE_URL.*planetscale", re.I), "planetscale"),
    (re.compile(r"UPSTASH_REDIS_URL|UPSTASH_REDIS_TOKEN", re.I), "upstash"),
    (re.compile(r"SENTRY_DSN|NEXT_PUBLIC_SENTRY", re.I), "sentry"),
    (re.compile(r"DATADOG_API_KEY|DD_API_KEY", re.I), "datadog"),
    (re.compile(r"NEW_RELIC_LICENSE_KEY", re.I), "new-relic"),
    (re.compile(r"LAUNCHDARKLY_SDK_KEY|LD_SDK_KEY", re.I), "launchdarkly"),
    (re.compile(r"HONEYCOMB_API_KEY", re.I), "honeycomb"),
    (re.compile(r"AXIOM_TOKEN|AXIOM_DATASET", re.I), "axiom"),
    (re.compile(r"CLERK_SECRET_KEY|NEXT_PUBLIC_CLERK", re.I), "clerk"),
    (re.compile(r"AUTH0_SECRET|AUTH0_DOMAIN", re.I), "auth0"),
    (re.compile(r"CONVEX_URL|CONVEX_DEPLOYMENT", re.I), "convex"),
    (re.compile(r"LANGFUSE_SECRET_KEY|LANGFUSE_PUBLIC_KEY", re.I), "langfuse"),
    (re.compile(r"CLOUDFLARE_API_TOKEN|CF_API_TOKEN", re.I), "cloudflare-workers"),
    (re.compile(r"FLY_API_TOKEN", re.I), "fly-io"),
    (re.compile(r"DOPPLER_TOKEN", re.I), "doppler"),
    (re.compile(r"INFISICAL_TOKEN", re.I), "infisical"),
    (re.compile(r"SNYK_TOKEN", re.I), "snyk"),
    (re.compile(r"FAUNA_SECRET", re.I), "fauna"),
]

# ── CLI Command Patterns ────────────────────────────────────────────

CLI_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r"\bsupabase\s+(?:init|start|link|db\s+push|migration)", re.I), "supabase"),
    (re.compile(r"\bneon\s+(?:databases|branches|connection-string)", re.I), "neon"),
    (re.compile(r"\bturso\s+(?:db|auth|group)", re.I), "turso"),
    (re.compile(r"\bwrangler\s+(?:dev|deploy|publish|init)", re.I), "cloudflare-workers"),
    (re.compile(r"\bflyctl\s+|fly\s+(?:deploy|launch|scale)", re.I), "fly-io"),
    (re.compile(r"\bdagger\s+(?:run|call)", re.I), "dagger"),
    (re.compile(r"\bvercel\s+(?:deploy|dev|env|link)", re.I), "vercel-edge-functions"),
    (re.compile(r"\bnetlify\s+(?:deploy|dev|init)", re.I), "netlify"),
    (re.compile(r"\bsentry-cli\s+", re.I), "sentry"),
    (re.compile(r"\bdoppler\s+(?:run|setup|secrets)", re.I), "doppler"),
    (re.compile(r"\binfisical\s+(?:init|run|secrets)", re.I), "infisical"),
]

# ── Context Phrases ─────────────────────────────────────────────────

COMPARISON_PHRASES = [
    re.compile(r"(?:compared?\s+to|versus|vs\.?|alternatively|or\s+you\s+could\s+use|another\s+option\s+is|instead\s+of)", re.I),
]

REJECTION_PHRASES = [
    re.compile(
        r"(?:(?:I\s+)?(?:wouldn't|would\s+not|don't|do\s+not)\s+recommend|avoid|stay\s+away\s+from"
        r"|not\s+(?:a\s+good|the\s+best)\s+(?:choice|option|fit)|(?:downside|drawback|limitation)s?\s+of)",
        re.I,
    ),
]

RECOMMENDATION_PHRASES = [
    re.compile(
        r"(?:I\s+(?:recommend|suggest)|you\s+(?:should|could)\s+use"
        r"|(?:great|good|excellent|best|ideal)\s+(?:choice|option|fit)|let's\s+use|we(?:'ll)?\s+use)",
        re.I,
    ),
]

IMPORT_PATTERN = re.compile(r"""(?:import\s+.*?from\s+['"]|require\s*\(\s*['"])([^'"]+)['"]""")


def extract_vendor_mentions(
    turn: ParsedTurn,
    taxonomy: VendorTaxonomy,
    user_prompt_snippet: str | None,
    *,
    package_resolver: Callable[[str], str | None] | None = None,
) -> list[VendorMention]:
    """Extract vendor mentions from a single turn (user or assistant).

    Returns a list of VendorMentions with deduplication within the turn.
    """
    resolve_package = package_resolver or _resolve_static
    mentions: list[VendorMention] = []
    seen: set[str] = set()

    def add_mention(
        vendor_canonical_id: str,
        vendor_raw: str,
        mention_type: str,
        confidence: float,
        context_snippet: str,
        timestamp: str,
    ) -> None:
        key = f"{vendor_canonical_id}:{mention_type}"
        if key in seen:
            return
        seen.add(key)

        entry = next((v for v in taxonomy.vendors if v.canonical_id == vendor_canonical_id), None)
        work_category = entry.category if entry else "other"

        mentions.append(
            VendorMention(
                vendor_canonical_id=vendor_canonical_id,
                vendor_raw=vendor_raw,
                mention_type=mention_type,  # type: ignore[arg-type]
                confidence=confidence,
                context_snippet=context_snippet,
                user_prompt_snippet=user_prompt_snippet,
                timestamp=timestamp,
                work_category=work_category,
            )
        )

    # ── Tier 1: Package install signals from tool_use blocks ────────
    for tool_use in turn.tool_uses:
        command = _get_command_from_tool_use(tool_use)
        if command:
            for pattern in INSTALL_PATTERNS:
                for match in pattern.finditer(command):
                    pkg_str = match.group(1)
                    pkgs = [p for p in pkg_str.split() if p and not p.startswith("-")]
                    for pkg in pkgs:
                        vendor_id = resolve_package(pkg)
                        if vendor_id:
                            add_mention(vendor_id, pkg, "installed", 1.0, command[:200], turn.timestamp)

            # Check CLI command patterns
            for pattern, vendor in CLI_PATTERNS:
                if pattern.search(command):
                    add_mention(vendor, command[:80], "configured", 0.9, command[:200], turn.timestamp)

    # ── Tier 2: Configuration signals from tool_use (Write/Edit) ────
    for tool_use in turn.tool_uses:
        content = _get_write_content_from_tool_use(tool_use)
        if not content:
            continue

        for pattern, vendor in CONNECTION_PATTERNS:
            if pattern.search(content):
                add_mention(vendor, pattern.pattern, "configured", 0.95, content[:200], turn.timestamp)

        for pattern, vendor in ENV_VAR_PATTERNS:
            if pattern.search(content):
                add_mention(vendor, pattern.pattern, "configured", 0.9, content[:200], turn.timestamp)

        # Check import statements → "implemented"
        for im in IMPORT_PATTERN.finditer(content):
            pkg = im.group(1)
            vendor_id = resolve_package(pkg)
            if vendor_id:
                add_mention(vendor_id, pkg, "implemented", 0.9, content[:200], turn.timestamp)

    # ── Tier 3: Text mentions from assistant text ─────────────────
    if turn.role == "assistant" and turn.text_content:
        _extract_text_mentions(turn.text_content, taxonomy, turn.timestamp, add_mention)

    return mentions


def _extract_text_mentions(
    text: str,
    taxonomy: VendorTaxonomy,
    timestamp: str,
    add_mention: Callable,
) -> None:
    """Extract vendor mentions from free text using taxonomy matching."""
    for vendor in taxonomy.vendors:
        terms = [vendor.display_name] + vendor.synonyms

        for term in terms:
            if len(term) < 3:
                continue

            escaped = re.escape(term)
            regex = re.compile(rf"\b{escaped}\b", re.I)
            match = regex.search(text)
            if not match:
                continue

            start_idx = max(0, match.start() - 100)
            end_idx = min(len(text), match.end() + 100)
            context = text[start_idx:end_idx]

            mention_type = "mentioned"
            confidence = 0.5

            if any(p.search(context) for p in REJECTION_PHRASES):
                mention_type = "rejected"
                confidence = 0.7
            elif any(p.search(context) for p in COMPARISON_PHRASES):
                mention_type = "compared"
                confidence = 0.7
            elif any(p.search(context) for p in RECOMMENDATION_PHRASES):
                mention_type = "recommended"
                confidence = 0.8

            add_mention(vendor.canonical_id, match.group(0), mention_type, confidence, context, timestamp)
            break  # Only match each vendor once per text block


def _get_command_from_tool_use(tool_use) -> str | None:
    """Extract command string from a tool_use record."""
    if tool_use.tool_name == "Bash" and isinstance(tool_use.input.get("command"), str):
        return tool_use.input["command"]
    if tool_use.tool_name in ("shell", "shell_command"):
        cmd = tool_use.input.get("command")
        if isinstance(cmd, list):
            return " ".join(str(c) for c in cmd)
        if isinstance(cmd, str):
            return cmd
        args = tool_use.input.get("arguments")
        if isinstance(args, str):
            try:
                parsed = json.loads(args)
                if isinstance(parsed.get("command"), list):
                    return " ".join(str(c) for c in parsed["command"])
            except (json.JSONDecodeError, TypeError):
                pass
    return None


def _get_write_content_from_tool_use(tool_use) -> str | None:
    """Extract written file content from a tool_use record."""
    if tool_use.tool_name in ("Write", "Edit"):
        content = tool_use.input.get("content")
        if isinstance(content, str):
            return content
        new_string = tool_use.input.get("new_string")
        if isinstance(new_string, str):
            return new_string
        return None
    if tool_use.tool_name == "apply_patch":
        raw = tool_use.input.get("raw")
        return raw if isinstance(raw, str) else None
    return None

File: packages/shared/src/extractor.ts
import type { VendorMention, MentionType, WorkCategory, VendorTaxonomy, ParsedTurn, UnknownPackage, ExtractionResult } from "./types.js";
import { resolvePackageToVendor as resolvePackageToVendorStatic, isBlocklistedPackage } from "./package-map.js";

/** Optional configuration for extractVendorMentions. */
export interface ExtractorOptions {
  /**
   * Custom package-name → vendor resolver. When provided, this is used instead
   * of the hardcoded PACKAGE_TO_VENDOR map.  Typically created via
   * `createPackageResolver(await loadPackageMapFromDb(pool))`.
   */
  packageResolver?: (packageName: string) => string | null;
}

// ── Package Manager Patterns ────────────────────────────────────────

const INSTALL_PATTERNS = [
  /(?:npm\s+install|npm\s+i|npx)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)/g,
  /(?:pnpm\s+add|pnpm\s+install)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)/g,
  /(?:yarn\s+add)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)/g,
  /(?:pip\s+install|pip3\s+install)\s+([^\s|&;]+(?:\s+[^\s|&;-][^\s|&;]*)*)/g,
  /(?:brew\s+install)\s+([^\s|&;]+)/g,
  /(?:docker\s+pull)\s+([^\s|&;]+)/g,
];

// ── Connection String Patterns ──────────────────────────────────────

const CONNECTION_PATTERNS: Array<{ pattern: RegExp; vendor: string }> = [
  { pattern: /neon\.tech/i, vendor: "neon" },
  { pattern: /supabase\.co/i, vendor: "supabase" },
  { pattern: /supabase\.com/i, vendor: "supabase" },
  { pattern: /planetscale\.com/i, vendor: "planetscale" },
  { pattern: /turso\.io/i, vendor: "turso" },
  { pattern: /cockroachlabs\.cloud/i, vendor: "cockroachdb" },
  { pattern: /upstash\.io/i, vendor: "upstash" },
  { pattern: /xata\.sh/i, vendor: "xata" },
  { pattern: /fauna\.com/i, vendor: "fauna" },
  { pattern: /convex\.cloud/i, vendor: "convex" },
  { pattern: /railway\.app/i, vendor: "railway-postgres" },
  { pattern: /vercel\.app/i, vendor: "vercel-edge-functions" },
  { pattern: /fly\.dev/i, vendor: "fly-io" },
  { pattern: /workers\.dev/i, vendor: "cloudflare-workers" },
  { pattern: /deno\.dev/i, vendor: "deno-deploy" },
];

// ── ENV Var Patterns ────────────────────────────────────────────────

const ENV_VAR_PATTERNS: Array<{ pattern: RegExp; vendor: string }> = [
  { pattern: /SUPABASE_URL|SUPABASE_KEY|SUPABASE_ANON_KEY|NEXT_PUBLIC_SUPABASE/i, vendor: "supabase" },
  { pattern: /NEON_DATABASE_URL|NEON_DB_URL/i, vendor: "neon" },
  { pattern: /TURSO_DATABASE_URL|TURSO_AUTH_TOKEN/i, vendor: "turso" },
  { pattern: /PLANETSCALE_|DATABASE_URL.*planetscale/i, vendor: "planetscale" },
  { pattern: /UPSTASH_REDIS_URL|UPSTASH_REDIS_TOKEN/i, vendor: "upstash" },
  { pattern: /SENTRY_DSN|NEXT_PUBLIC_SENTRY/i, vendor: "sentry" },
  { pattern: /DATADOG_API_KEY|DD_API_KEY/i, vendor: "datadog" },
  { pattern: /NEW_RELIC_LICENSE_KEY/i, vendor: "new-relic" },
  { pattern: /LAUNCHDARKLY_SDK_KEY|LD_SDK_KEY/i, vendor: "launchdarkly" },
  { pattern: /HONEYCOMB_API_KEY/i, vendor: "honeycomb" },
  { pattern: /AXIOM_TOKEN|AXIOM_DATASET/i, vendor: "axiom" },
  { pattern: /CLERK_SECRET_KEY|NEXT_PUBLIC_CLERK/i, vendor: "clerk" },
  { pattern: /AUTH0_SECRET|AUTH0_DOMAIN/i, vendor: "auth0" },
  { pattern: /CONVEX_URL|CONVEX_DEPLOYMENT/i, vendor: "convex" },
  { pattern: /LANGFUSE_SECRET_KEY|LANGFUSE_PUBLIC_KEY/i, vendor: "langfuse" },
  { pattern: /CLOUDFLARE_API_TOKEN|CF_API_TOKEN/i, vendor: "cloudflare-workers" },
  { pattern: /FLY_API_TOKEN/i, vendor: "fly-io" },
  { pattern: /DOPPLER_TOKEN/i, vendor: "doppler" },
  { pattern: /INFISICAL_TOKEN/i, vendor: "infisical" },
  { pattern: /SNYK_TOKEN/i, vendor: "snyk" },
  { pattern: /FAUNA_SECRET/i, vendor: "fauna" },
];

// ── CLI Command Patterns ────────────────────────────────────────────

const CLI_PATTERNS: Array<{ pattern: RegExp; vendor: string }> = [
  { pattern: /\bsupabase\s+(?:init|start|link|db\s+push|migration)/i, vendor: "supabase" },
  { pattern: /\bneon\s+(?:databases|branches|connection-string)/i, vendor: "neon" },
  { pattern: /\bturso\s+(?:db|auth|group)/i, vendor: "turso" },
  { pattern: /\bwrangler\s+(?:dev|deploy|publish|init)/i, vendor: "cloudflare-workers" },
  { pattern: /\bflyctl\s+|fly\s+(?:deploy|launch|scale)/i, vendor: "fly-io" },
  { pattern: /\bdagger\s+(?:run|call)/i, vendor: "dagger" },
  { pattern: /\bvercel\s+(?:deploy|dev|env|link)/i, vendor: "vercel-edge-functions" },
  { pattern: /\bnetlify\s+(?:deploy|dev|init)/i, vendor: "netlify" },
  { pattern: /\bsentry-cli\s+/i, vendor: "sentry" },
  { pattern: /\bdoppler\s+(?:run|setup|secrets)/i, vendor: "doppler" },
  { pattern: /\binfisical\s+(?:init|run|secrets)/i, vendor: "infisical" },
];

// ── Comparison / Rejection Phrases ──────────────────────────────────

const COMPARISON_PHRASES = [
  /(?:compared?\s+to|versus|vs\.?|alternatively|or\s+you\s+could\s+use|another\s+option\s+is|instead\s+of)/i,
];

const REJECTION_PHRASES = [
  /(?:(?:I\s+)?(?:wouldn't|would\s+not|don't|do\s+not)\s+recommend|avoid|stay\s+away\s+from|not\s+(?:a\s+good|the\s+best)\s+(?:choice|option|fit)|(?:downside|drawback|limitation)s?\s+of)/i,
];

const RECOMMENDATION_PHRASES = [
  /(?:I\s+(?:recommend|suggest)|you\s+(?:should|could)\s+use|(?:great|good|excellent|best|ideal)\s+(?:choice|option|fit)|let's\s+use|we(?:'ll)?\s+use)/i,
];

// ── Main Extraction Function ────────────────────────────────────────

/**
 * Extract vendor mentions from a single turn (user or assistant).
 * Returns an array of VendorMentions with deduplication within the turn.
 *
 * @param options.packageResolver - custom resolver for package → vendor mapping.
 *   When omitted, falls back to the static PACKAGE_TO_VENDOR map.
 */
export function extractVendorMentions(
  turn: ParsedTurn,
  taxonomy: VendorTaxonomy,
  userPromptSnippet: string | null,
  options?: ExtractorOptions,
): VendorMention[] {
  const resolvePackage = options?.packageResolver ?? resolvePackageToVendorStatic;
  const mentions: VendorMention[] = [];
  const seen = new Set<string>(); // "vendorId:mentionType" for dedup within turn

  function addMention(m: Omit<VendorMention, "userPromptSnippet">) {
    const key = `${m.vendorCanonicalId}:${m.mentionType}`;
    if (seen.has(key)) return;
    seen.add(key);

    // Look up work category from taxonomy
    const entry = taxonomy.vendors.find(v => v.canonical_id === m.vendorCanonicalId);
    const workCategory = (entry?.category as WorkCategory) ?? "other";

    mentions.push({ ...m, userPromptSnippet, workCategory });
  }

  // ── Tier 1: Package install signals from tool_use blocks ────────
  for (const toolUse of turn.toolUses) {
    const command = getCommandFromToolUse(toolUse);
    if (!command) continue;

    for (const pattern of INSTALL_PATTERNS) {
      pattern.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = pattern.exec(command)) !== null) {
        const pkgStr = match[1];
        // Split on whitespace to handle "npm install pkg1 pkg2"
        const pkgs = pkgStr.split(/\s+/).filter(p => p && !p.startsWith("-"));
        for (const pkg of pkgs) {
          const vendorId = resolvePackage(pkg);
          if (vendorId) {
            addMention({
              vendorCanonicalId: vendorId,
              vendorRaw: pkg,
              mentionType: "installed",
              confidence: 1.0,
              contextSnippet: command.slice(0, 200),
              timestamp: turn.timestamp,
              workCategory: null,
            });
          }
        }
      }
    }

    // Check for CLI command patterns
    if (command) {
      for (const { pattern, vendor } of CLI_PATTERNS) {
        if (pattern.test(command)) {
          addMention({
            vendorCanonicalId: vendor,
            vendorRaw: command.slice(0, 80),
            mentionType: "configured",
            confidence: 0.9,
            contextSnippet: command.slice(0, 200),
            timestamp: turn.timestamp,
            workCategory: null,
          });
        }
      }
    }
  }

  // ── Tier 2: Configuration signals from tool_use (Write/Edit) ────
  for (const toolUse of turn.toolUses) {
    const content = getWriteContentFromToolUse(toolUse);
    if (!content) continue;

    // Check connection strings
    for (const { pattern, vendor } of CONNECTION_PATTERNS) {
      if (pattern.test(content)) {
        addMention({
          vendorCanonicalId: vendor,
          vendorRaw: pattern.source,
          mentionType: "configured",
          confidence: 0.95,
          contextSnippet: content.slice(0, 200),
          timestamp: turn.timestamp,
          workCategory: null,
        });
      }
    }

    // Check env vars
    for (const { pattern, vendor } of ENV_VAR_PATTERNS) {
      if (pattern.test(content)) {
        addMention({
          vendorCanonicalId: vendor,
          vendorRaw: pattern.source,
          mentionType: "configured",
          confidence: 0.9,
          contextSnippet: content.slice(0, 200),
          timestamp: turn.timestamp,
          workCategory: null,
        });
      }
    }

    // Check import statements → "implemented"
    const importMatches = content.matchAll(
      /(?:import\s+.*?from\s+['"]|require\s*\(\s*['"])([^'"]+)['"]/g,
    );
    for (const im of importMatches) {
      const pkg = im[1];
      const vendorId = resolvePackage(pkg);
      if (vendorId) {
        addMention({
          vendorCanonicalId: vendorId,
          vendorRaw: pkg,
          mentionType: "implemented",
          confidence: 0.9,
          contextSnippet: content.slice(0, 200),
          timestamp: turn.timestamp,
          workCategory: null,
        });
      }
    }
  }

  // ── Tier 3: Text mentions from assistant text ─────────────────
  if (turn.role === "assistant" && turn.textContent) {
    extractTextMentions(turn.textContent, taxonomy, turn.timestamp, addMention);
  }

  return mentions;
}

/**
 * Extract vendor mentions AND collect unknown packages in one pass.
 * Unknown packages are those that appear in install commands or import
 * statements but don't resolve to any known vendor and aren't blocklisted.
 *
 * The caller (ingest-bridge) uses the unknownPackages list to create
 * dynamic vendor entries, then re-extracts to catch text mentions.
 */
export function extractVendorMentionsWithUnknowns(
  turn: ParsedTurn,
  taxonomy: VendorTaxonomy,
  userPromptSnippet: string | null,
  options?: ExtractorOptions,
): ExtractionResult {
  const resolvePackage = options?.packageResolver ?? resolvePackageToVendorStatic;
  const mentions: VendorMention[] = [];
  const unknownPackages: UnknownPackage[] = [];
  const seen = new Set<string>();
  const seenUnknown = new Set<string>();

  function addMention(m: Omit<VendorMention, "userPromptSnippet">) {
    const key = `${m.vendorCanonicalId}:${m.mentionType}`;
    if (seen.has(key)) return;
    seen.add(key);
    const entry = taxonomy.vendors.find(v => v.canonical_id === m.vendorCanonicalId);
    const workCategory = (entry?.category as WorkCategory) ?? "other";
    mentions.push({ ...m, userPromptSnippet, workCategory });
  }

  function addUnknown(pkg: string, command: string, timestamp: string) {
    if (seenUnknown.has(pkg)) return;
    seenUnknown.add(pkg);
    unknownPackages.push({
      packageName: pkg,
      installCommand: command.slice(0, 300),
      timestamp,
      contextSnippet: command.slice(0, 200),
    });
  }

  // ── Tier 1: Package install signals from tool_use blocks ────────
  for (const toolUse of turn.toolUses) {
    const command = getCommandFromToolUse(toolUse);
    if (!command) continue;

    for (const pattern of INSTALL_PATTERNS) {
      pattern.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = pattern.exec(command)) !== null) {
        const pkgStr = match[1];
        const pkgs = pkgStr.split(/\s+/).filter(p => p && !p.startsWith("-"));
        for (const pkg of pkgs) {
          const vendorId = resolvePackage(pkg);
          if (vendorId) {
            addMention({
              vendorCanonicalId: vendorId,
              vendorRaw: pkg,
              mentionType: "installed",
              confidence: 1.0,
              contextSnippet: command.slice(0, 200),
              timestamp: turn.timestamp,
              workCategory: null,
            });
          } else {
            const cleaned = pkg.replace(/@[\d^~>=<.*]+$/, "");
            if (!isBlocklistedPackage(cleaned)) {
              addUnknown(cleaned, command, turn.timestamp);
            }
          }
        }
      }
    }

    // CLI command patterns (same as original — no unknown collection needed)
    if (command) {
      for (const { pattern, vendor } of CLI_PATTERNS) {
        if (pattern.test(command)) {
          addMention({
            vendorCanonicalId: vendor,
            vendorRaw: command.slice(0, 80),
            mentionType: "configured",
            confidence: 0.9,
            contextSnippet: command.slice(0, 200),
            timestamp: turn.timestamp,
            workCategory: null,
          });
        }
      }
    }
  }

  // ── Tier 2: Configuration signals from tool_use (Write/Edit) ────
  for (const toolUse of turn.toolUses) {
    const content = getWriteContentFromToolUse(toolUse);
    if (!content) continue;

    for (const { pattern, vendor } of CONNECTION_PATTERNS) {
      if (pattern.test(content)) {
        addMention({
          vendorCanonicalId: vendor,
          vendorRaw: pattern.source,
          mentionType: "configured",
          confidence: 0.95,
          contextSnippet: content.slice(0, 200),
          timestamp: turn.timestamp,
          workCategory: null,
        });
      }
    }

    for (const { pattern, vendor } of ENV_VAR_PATTERNS) {
      if (pattern.test(content)) {
        addMention({
          vendorCanonicalId: vendor,
          vendorRaw: pattern.source,
          mentionType: "configured",
          confidence: 0.9,
          contextSnippet: content.slice(0, 200),
          timestamp: turn.timestamp,
          workCategory: null,
        });
      }
    }

    // Import statements — collect unknowns here too
    const importMatches = content.matchAll(
      /(?:import\s+.*?from\s+['"]|require\s*\(\s*['"])([^'"]+)['"]/g,
    );
    for (const im of importMatches) {
      const pkg = im[1];
      // Skip relative imports
      if (pkg.startsWith(".") || pkg.startsWith("/")) continue;
      const vendorId = resolvePackage(pkg);
      if (vendorId) {
        addMention({
          vendorCanonicalId: vendorId,
          vendorRaw: pkg,
          mentionType: "implemented",
          confidence: 0.9,
          contextSnippet: content.slice(0, 200),
          timestamp: turn.timestamp,
          workCategory: null,
        });
      } else {
        const cleaned = pkg.replace(/@[\d^~>=<.*]+$/, "");
        if (!isBlocklistedPackage(cleaned)) {
          addUnknown(cleaned, `import ${pkg}`, turn.timestamp);
        }
      }
    }
  }

  // ── Tier 3: Text mentions from assistant text ─────────────────
  if (turn.role === "assistant" && turn.textContent) {
    extractTextMentions(turn.textContent, taxonomy, turn.timestamp, addMention);
  }

  return { mentions, unknownPackages };
}

/**
 * Extract vendor mentions from free text using taxonomy matching.
 */
function extractTextMentions(
  text: string,
  taxonomy: VendorTaxonomy,
  timestamp: string,
  addMention: (m: Omit<VendorMention, "userPromptSnippet">) => void,
): void {
  for (const vendor of taxonomy.vendors) {
    // Build list of terms to match
    const terms = [vendor.display_name, ...vendor.synonyms];

    for (const term of terms) {
      // Skip very short terms that cause false positives
      if (term.length < 3) continue;

      // Word-boundary match (case-insensitive)
      const escaped = term.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      const regex = new RegExp(`\\b${escaped}\\b`, "i");
      const match = regex.exec(text);

      if (!match) continue;

      // Determine mention type based on surrounding context
      const startIdx = Math.max(0, match.index - 100);
      const endIdx = Math.min(text.length, match.index + term.length + 100);
      const context = text.slice(startIdx, endIdx);

      let mentionType: MentionType = "mentioned";
      let confidence = 0.5;

      // Check for rejection context
      if (REJECTION_PHRASES.some(p => p.test(context))) {
        mentionType = "rejected";
        confidence = 0.7;
      }
      // Check for comparison context
      else if (COMPARISON_PHRASES.some(p => p.test(context))) {
        mentionType = "compared";
        confidence = 0.7;
      }
      // Check for recommendation context
      else if (RECOMMENDATION_PHRASES.some(p => p.test(context))) {
        mentionType = "recommended";
        confidence = 0.8;
      }

      addMention({
        vendorCanonicalId: vendor.canonical_id,
        vendorRaw: match[0],
        mentionType,
        confidence,
        contextSnippet: context,
        timestamp,
        workCategory: null,
      });

      break; // Only match each vendor once per text block
    }
  }
}

// ── Helper: Extract command from tool_use ────────────────────────────

function getCommandFromToolUse(toolUse: { toolName: string; input: Record<string, unknown> }): string | null {
  // Claude Code: Bash tool with "command" input
  if (toolUse.toolName === "Bash" && typeof toolUse.input.command === "string") {
    return toolUse.input.command;
  }
  // Codex CLI: shell/shell_command function with "command" input
  if (toolUse.toolName === "shell" || toolUse.toolName === "shell_command") {
    const cmd = toolUse.input.command;
    if (Array.isArray(cmd)) {
      return (cmd as string[]).join(" ");
    }
    if (typeof cmd === "string") return cmd;

    // May be in arguments as JSON string
    const args = toolUse.input.arguments;
    if (typeof args === "string") {
      try {
        const parsed = JSON.parse(args);
        if (Array.isArray(parsed.command)) return parsed.command.join(" ");
      } catch { /* ignore */ }
    }
  }
  return null;
}

function getWriteContentFromToolUse(toolUse: { toolName: string; input: Record<string, unknown> }): string | null {
  if (toolUse.toolName === "Write" || toolUse.toolName === "Edit") {
    return (
      (typeof toolUse.input.content === "string" ? toolUse.input.content : null) ??
      (typeof toolUse.input.new_string === "string" ? toolUse.input.new_string : null)
    );
  }
  // Codex CLI: apply_patch tool contains file content as raw string
  if (toolUse.toolName === "apply_patch") {
    return typeof toolUse.input.raw === "string" ? toolUse.input.raw : null;
  }
  return null;
}
